crates/autopilot/src/run_loop.rs

```diff
@@ -74,14 +83,88 @@ impl RunLoop {
+            if solution.score == 0.into() {
                 return;
             }
+
+            let auction_id = id;
+            let winner = solution.submission_address;
+            let winning_score = solution.score;
+            let reference_score = solutions
+                .last()
+                .map(|(_, response)| response.score)
+                .unwrap_or_default();
+            let mut participants = solutions
+                .iter()
+                .map(|(_, response)| response.submission_address)
+                .collect::<HashSet<_>>();
+            participants.insert(solution.submission_address); // add winner as participant
+
+            let mut prices = BTreeMap::new();
+            let block_deadline = self.current_block.borrow().number
+                + self.submission_deadline
+                + self.additional_deadline_for_rewards;
+            // save order executions for all orders in the solution, except for partial
+            // limit orders. order executions for partial limit orders will be
+            // saved after settling the order onchain.
+            let mut order_executions = vec![];
+            for order_id in &solution.orders {
+                let auction_order = auction
+                    .orders
+                    .iter()
+                    .find(|auction_order| &auction_order.metadata.uid == order_id);
+                match auction_order {
+                    Some(auction_order) => {
+                        let executed_fee = match auction_order.metadata.class {
+                            OrderClass::Limit(LimitOrderClass { surplus_fee, .. }) => {
+                                match auction_order.data.partially_fillable {
+                                    // we don't know the surplus fee in advance. will be populated
+                                    // after the transaction containing the order is mined
+                                    true => ExecutedFee::Surplus(None),
+                                    // calculated by the protocol, therefore we can trust the
+                                    // auction
+                                    false => ExecutedFee::Surplus(surplus_fee),
+                                }
+                            }
+                            _ => ExecutedFee::Solver(auction_order.metadata.solver_fee),
+                        };
+                        order_executions.push(OrderExecution {
+                            order_id: *order_id,
+                            executed_fee,
+                        });
+                        if let Some(price) = auction.prices.get(&auction_order.data.sell_token) {
+                            prices.insert(auction_order.data.sell_token, *price);
+                        }
+                        if let Some(price) = auction.prices.get(&auction_order.data.buy_token) {
+                            prices.insert(auction_order.data.buy_token, *price);
+                        }
```

Comment:
Can we log an `ERROR` if the token is missing? It should really *never* happen, and if it does, something is sufficiently wrong that it should be investigated.